config.go

```diff
@@ -304,7 +304,7 @@ func loadConfig() (*config, error) {
 
 	// Pre-parse the command line options to pick up an alternative config
 	// file.
-	preCfg := defaultCfg
+	preCfg := config{}
```

Comment:
This can be left as is, the only change required is below.

Reply:
Per the description of the algorithm, the first step is to pre-parse the command line to check for an alternative config file. So the simplest way to know that the user added --lnddir and/or --configfile is to initialize preCfg to an empty struct then compare these two values with "". If we initialize preCfg to defaultCfg, we have no simple way to know if the user didn't set set the value the same as the default value or didn't set the value.

And the two following cases are differents:
lnd --lnddir ~/.lnd2 --configfile ~/.lnd/lnd.conf
lnd --lnddir ~/.lnd2
In the first case, the config file is ~/.lnd/lnd.conf and in the second case the config file is ~/.lnd2/lnd.conf

Reply:
The advantage of using defaultCfg instead of config{} is when the --help flag is used for instance, because it gives the ability to go-flags to show the default values. So the new code below adds another parsing phase woth the default value when there is an error

Reply:
@yaslama, I think we can work around this with the following diff, let me know what you think.
```
diff --git a/config.go b/config.go
index 28118661..52762f09 100644
--- a/config.go
+++ b/config.go
@@ -322,15 +322,15 @@ func loadConfig() (*config, error) {
 	// path to all of the files and directories that will live within it.
 	lndDir := cleanAndExpandPath(preCfg.LndDir)
 	if lndDir != defaultLndDir {
-		defaultCfg.ConfigFile = filepath.Join(lndDir, defaultConfigFilename)
-		defaultCfg.DataDir = filepath.Join(lndDir, defaultDataDirname)
-		defaultCfg.TLSCertPath = filepath.Join(lndDir, defaultTLSCertFilename)
-		defaultCfg.TLSKeyPath = filepath.Join(lndDir, defaultTLSKeyFilename)
-		defaultCfg.AdminMacPath = filepath.Join(lndDir, defaultAdminMacFilename)
-		defaultCfg.InvoiceMacPath = filepath.Join(lndDir, defaultInvoiceMacFilename)
-		defaultCfg.ReadMacPath = filepath.Join(lndDir, defaultReadMacFilename)
-		defaultCfg.LogDir = filepath.Join(lndDir, defaultLogDirname)
-		defaultCfg.Tor.V2PrivateKeyPath = filepath.Join(lndDir, defaultTorV2PrivateKeyFilename)
+		preCfg.ConfigFile = filepath.Join(lndDir, defaultConfigFilename)
+		preCfg.DataDir = filepath.Join(lndDir, defaultDataDirname)
+		preCfg.TLSCertPath = filepath.Join(lndDir, defaultTLSCertFilename)
+		preCfg.TLSKeyPath = filepath.Join(lndDir, defaultTLSKeyFilename)
+		preCfg.AdminMacPath = filepath.Join(lndDir, defaultAdminMacFilename)
+		preCfg.InvoiceMacPath = filepath.Join(lndDir, defaultInvoiceMacFilename)
+		preCfg.ReadMacPath = filepath.Join(lndDir, defaultReadMacFilename)
+		preCfg.LogDir = filepath.Join(lndDir, defaultLogDirname)
+		preCfg.Tor.V2PrivateKeyPath = filepath.Join(lndDir, defaultTorV2PrivateKeyFilename)
 	}
 
 	// Create the lnd directory if it doesn't already exist.
@@ -354,9 +354,8 @@ func loadConfig() (*config, error) {
 
 	// Next, load any additional configuration options from the file.
 	var configFileError error
-	cfg := defaultCfg
-	configFile := cleanAndExpandPath(preCfg.ConfigFile)
-	if err := flags.IniParse(configFile, &cfg); err != nil {
+	cfg := preCfg
+	if err := flags.IniParse(preCfg.ConfigFile, &cfg); err != nil {
 		configFileError = err
 	}
 
@@ -366,6 +365,15 @@ func loadConfig() (*config, error) {
 		return nil, err
 	}
 
+	// If it's the case that the path to the config file has changed once we
+	// we've reparsed the command line options, we'll parse the new config
+	// file specified.
+	if cfg.ConfigFile != preCfg.ConfigFile {
+		if err := flags.IniParse(cfg.ConfigFile, &cfg); err != nil {
+			configFileError = err
+		}
+	}
+
 	// As soon as we're done parsing configuration options, ensure all paths
 	// to directories and files are cleaned and expanded before attempting
 	// to use them later on.
```

Reply:
@wpaulino Thanks for your comment and your proposition. The problem with this patch is that one superfluous wrong ini file will be parsed and can contain values which are not the default values and are not in the good ini file.
For instance, if the user runs lnd as:

``lnd --lnddir=~/.lnd2 --configfile=~/.lnd/lnd.conf``

when ~/.lnd2/lnd.conf contains nolisten=1 but not ~/.lnd/lnd.conf

Then lnd will not listen to other peers which is not what the user wants.

Reply:
Good point! I believe this should handle that now.
```
diff --git a/config.go b/config.go
index 28118661..10bab1c9 100644
--- a/config.go
+++ b/config.go
@@ -322,15 +322,21 @@ func loadConfig() (*config, error) {
 	// path to all of the files and directories that will live within it.
 	lndDir := cleanAndExpandPath(preCfg.LndDir)
 	if lndDir != defaultLndDir {
-		defaultCfg.ConfigFile = filepath.Join(lndDir, defaultConfigFilename)
-		defaultCfg.DataDir = filepath.Join(lndDir, defaultDataDirname)
-		defaultCfg.TLSCertPath = filepath.Join(lndDir, defaultTLSCertFilename)
-		defaultCfg.TLSKeyPath = filepath.Join(lndDir, defaultTLSKeyFilename)
-		defaultCfg.AdminMacPath = filepath.Join(lndDir, defaultAdminMacFilename)
-		defaultCfg.InvoiceMacPath = filepath.Join(lndDir, defaultInvoiceMacFilename)
-		defaultCfg.ReadMacPath = filepath.Join(lndDir, defaultReadMacFilename)
-		defaultCfg.LogDir = filepath.Join(lndDir, defaultLogDirname)
-		defaultCfg.Tor.V2PrivateKeyPath = filepath.Join(lndDir, defaultTorV2PrivateKeyFilename)
+		// We'll only modify the path to the config file if the user has
+		// not explicitly set one. Otherwise, we'll ignore this setting
+		// and use the config file provided.
+		if preCfg.ConfigFile == defaultConfigFile {
+			preCfg.ConfigFile = filepath.Join(lndDir, defaultConfigFilename)
+		}
+
+		preCfg.DataDir = filepath.Join(lndDir, defaultDataDirname)
+		preCfg.TLSCertPath = filepath.Join(lndDir, defaultTLSCertFilename)
+		preCfg.TLSKeyPath = filepath.Join(lndDir, defaultTLSKeyFilename)
+		preCfg.AdminMacPath = filepath.Join(lndDir, defaultAdminMacFilename)
+		preCfg.InvoiceMacPath = filepath.Join(lndDir, defaultInvoiceMacFilename)
+		preCfg.ReadMacPath = filepath.Join(lndDir, defaultReadMacFilename)
+		preCfg.LogDir = filepath.Join(lndDir, defaultLogDirname)
+		preCfg.Tor.V2PrivateKeyPath = filepath.Join(lndDir, defaultTorV2PrivateKeyFilename)
 	}
 
 	// Create the lnd directory if it doesn't already exist.
@@ -354,9 +360,8 @@ func loadConfig() (*config, error) {
 
 	// Next, load any additional configuration options from the file.
 	var configFileError error
-	cfg := defaultCfg
-	configFile := cleanAndExpandPath(preCfg.ConfigFile)
-	if err := flags.IniParse(configFile, &cfg); err != nil {
+	cfg := preCfg
+	if err := flags.IniParse(preCfg.ConfigFile, &cfg); err != nil {
 		configFileError = err
 	}

```

Reply:
The problem is that in the following two cases:
1) lnd --lnddir ~/.lnd2 --configfile ~/.lnd/lnd.conf
2) lnd --lnddir ~/.lnd2

We have lndDir != defaultLndDir and preCfg.ConfigFile == defaultConfigFile so with your code, we will have ~/.lnd2/lnd.conf in preCfg.ConfigFile and this file will be parsed.
It's ok for the case 2 but not for the case 1.

Reply:
The latest diff does handle the first case now as well. I've tested this locally to confirm.

Reply:
I made a mistake in my example. The problem occurs only if you use full path because defaultConfigFile is a full path.
I tried (my home directory is /home/yaslama) using your last diff:
``lnd --lnddir=/home/yaslama/.lnd2 --configfile=/home/yaslama/.lnd/lnd.conf``
and the ini file which was parsed was /home/yaslama/.lnd2/lnd.conf

Reply:
I'm not able to reproduce this with the diff above.

`lnd --lnddir=~/test` *only* picks up the config file at `~/test/lnd.conf`.

`lnd --lnddir=~/test --configfile=~/lnd.conf` *only* picks up the config file at `~/lnd.conf`.

Reply:
@wpaulino Sorry to insist, but can you try your new diff above using the following cases:
``lnd --lnddir=~/test --configfile=~/.lnd/lnd.conf``
``lnd --lnddir=~/test``
In both cases it will pick ~/test/lnd.conf
You used in your second test:
``--configfile=~/lnd.conf``
instead of
``--configfile=~/.lnd/lnd.conf``
The later value is the default, not the former.

Reply:
Are you applying this diff on your branch? It's meant to be for master, sorry if that wasn't clear.

Reply:
Yes. I am applying the diff on master. Did you try with ~/.lnd/lnd.conf and not ~/lnd.conf which is not default?

Reply:
@wpaulino and @Roasbeef, I simplified the diff in #1451 
Can you please check?